src/picongpu/include/plugins/particleMerging/VoronoiCell.hpp

```diff
@@ -0,0 +1,261 @@
+    struct VoronoiCell
+    {
+        VoronoiStatus status;
+        VoronoiSplittingStage splittingStage;
+        uint32_t numMacroParticles;
+        float_X numRealParticles;
+
+        float3_X meanValue;
+        float3_X meanSquaredValue;
+
+        uint8_t splittingComponent;
+        int32_t lowerCellId;
+        int32_t higherCellId;
+        int firstParticleFlag;
+
+        HDINLINE
+        VoronoiCell( VoronoiSplittingStage splittingStage = VoronoiSplittingStage::position ) :
+            status( VoronoiStatus::collecting ),
+            splittingStage( splittingStage ),
+            numMacroParticles( 0 ),
+            numRealParticles( float_X( 0.0 ) ),
+            meanValue( float3_X::create( 0.0 ) ),
+            meanSquaredValue( float3_X::create( 0.0 ) ),
+            firstParticleFlag( 0 )
+        {}
+
+
+        HDINLINE
+        void setToAbort()
+        {
+            this->status = VoronoiStatus::abort;
+        }
+
+
+        HDINLINE
+        void setToSplitting(
+            const uint8_t splittingComponent,
+            const int32_t lowerCellId,
+            const int32_t higherCellId)
+        {
+            this->status = VoronoiStatus::splitting;
+            this->splittingComponent = splittingComponent;
+            this->lowerCellId = lowerCellId;
+            this->higherCellId = higherCellId;
+        }
+
+
+        HDINLINE
+        void setToReadyForMerging()
+        {
+            this->status = VoronoiStatus::readyForMerging;
+        }
+
+        DINLINE
+        bool isFirstParticle()
+        {
+            return atomicExch( &this->firstParticleFlag, 1 ) == 0;
+        }
+
+
+        DINLINE
+        void addParticle(
+            const floatD_X position,
+            const float3_X momentum,
+            const float_X weighting
+        )
+        {
+            nvidia::atomicAllInc( &this->numMacroParticles );
+            nvidia::atomicAdd( &this->numRealParticles, weighting );
+
+            if( this->splittingStage == VoronoiSplittingStage::position )
+            {
+                const floatD_X position2 = position * position;
+
+                for( int i = 0; i < simDim; i++ )
+                {
+                    nvidia::atomicAdd( &this->meanValue[i], weighting * position[i] );
+                    nvidia::atomicAdd( &this->meanSquaredValue[i], weighting * position2[i] );
+                }
+            }
+            else
+            {
+                const float3_X momentum2 = momentum * momentum;
+
+                for( int i = 0; i < DIM3; i++ )
+                {
+                    nvidia::atomicAdd( &this->meanValue[i], weighting * momentum[i] );
+                    nvidia::atomicAdd( &this->meanSquaredValue[i], weighting * momentum2[i] );
+                }
+            }
+        }
+
+
+        /** finalize mean value calculation */
+        HDINLINE
+        void finalizeMeanValues()
+        {
+            this->meanValue /= this->numRealParticles;
```

Comment:
In [this line](https://github.com/ComputationalRadiationPhysics/picongpu/pull/1959/files#diff-00f205db21c4990b088ff306527f100fR143) `meanValue` is changes by all threads with a atomic function. If a warp is still changing `meanValue` during a other thread is performing `finalizeMeanValues()` than we have a classical race condition.